src/timeout.rs

```diff
@@ -1,29 +1,181 @@
+use http::StatusCode;
+use std::sync::Arc;
 use std::time::Duration;
 use tokio::time;
 
 use crate::middleware::{BoxFuture, Middleware};
-use crate::{Next, Request};
+use crate::{Error, Next, Request};
 
+/// Require dowstream middleware to respond within a specified duration.
+///
+/// If the timeout expires before a response is generated, an error is returned.
+///
+/// # Example
+///
+/// ```
+/// use std::time::Duration;
+/// use tokio::time::sleep;
+/// use via::{App, Response, Timeout};
+///
+/// let mut app = App::new(());
+///
+/// app.middleware(Timeout::new(Duration::from_secs(10)));
+/// app.route("/").respond(via::get(async |_, _| {
+///     sleep(Duration::from_secs(11)).await;
+///     Response::build().text("Hello, world!")
+/// }));
+/// ```
+///
 pub struct Timeout {
     duration: Duration,
+    or_else: OrElse,
 }
 
-/// Create a new `Timeout` middleware with the specified duration.
-pub fn timeout(duration: Duration) -> Timeout {
-    Timeout { duration }
+enum OrElse {
```

Comment:
Implemented as an enum to enforce mutual exclusivity in the type system.

Otherwise, we'd have to override the status code of the response returned from the or_else callback which is redundant if a user is expected to return a result that contains a Response.